sql/catalyst/src/main/java/org/apache/spark/sql/connector/CustomMetric.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector;
+
+import org.apache.spark.annotation.Evolving;
+
+/**
+ * A custom metric.
```

Comment:
Since this is a public API, it would be great to add more information to explain how to use APIs as a developer. For example,

- If I define my own `CustomMetric`, how does Spark use it?
- If I define a metric type that doesn't support `sum`, for example, measure the executor jvm heap size, how does Spark handle it?
- If my `PartitionReaderWithMetrics` returns metrics for a partition, will Spark combine them for partitions of a Spark job?
- In the streaming case, how does Spark combine metrics from different micro batches?
- If I would like to report metrics in driver, how do I do it?

I feel you might need to build an interface similar to the private `SQLMetric`.

Reply:
I will add more information to the comment.

Basically I will add a few metric classes based on `CustomMetric`. They are correspond to sum, size, timing metrics in `SQLMetric`.